File: qt/src/services/RedesService.cpp
#include "services/RedesService.h"

#include "db/Database.h"

#include <QClipboard>
#include <QDateTime>
#include <QGuiApplication>
#include <QSqlQuery>
#include <QVariant>

const QString RedesService::kSelect = QStringLiteral(
    "SELECT id, titulo, contenido, emojis, hashtags, tipo, "
    "foto_ids AS fotoIds, fecha_creacion AS fechaCreacion, "
    "fecha_programada AS fechaProgramada, publicado, plataforma, "
    "visualizaciones, notas FROM posts_redes ");

RedesService::RedesService(QObject *parent) : QObject(parent) {}

QVariantList RedesService::obtenerTodos() const
{
    QSqlQuery q = Database::instance().exec(
        kSelect + QStringLiteral("ORDER BY fecha_creacion DESC"));
    return Database::rows(q);
}

QVariantList RedesService::filtrar(const QString &filtro) const
{
    QString where;
    if (filtro == QStringLiteral("pendientes"))
        where = QStringLiteral("WHERE publicado = 0 ");
    else if (filtro == QStringLiteral("publicados"))
        where = QStringLiteral("WHERE publicado = 1 ");
    QSqlQuery q = Database::instance().exec(
        kSelect + where + QStringLiteral("ORDER BY fecha_creacion DESC"));
    return Database::rows(q);
}

QVariantMap RedesService::obtenerPorId(int id) const
{
    QSqlQuery q = Database::instance().exec(kSelect + QStringLiteral("WHERE id = ?"), {id});
    return Database::firstRow(q);
}

int RedesService::crear(const QVariantMap &datos)
{
    QString fecha = datos.value(QStringLiteral("fechaCreacion")).toString();
    if (fecha.isEmpty())
        fecha = QDateTime::currentDateTime().toString(Qt::ISODate);
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("INSERT INTO posts_redes (titulo, contenido, emojis, hashtags, tipo, "
                       "foto_ids, fecha_creacion, fecha_programada, publicado, plataforma, "
                       "visualizaciones, notas) VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0, ?, 0, ?)"),
        {datos.value(QStringLiteral("titulo")),
         datos.value(QStringLiteral("contenido")),
         datos.value(QStringLiteral("emojis")),
         datos.value(QStringLiteral("hashtags")),
         datos.value(QStringLiteral("tipo")),
         datos.value(QStringLiteral("fotoIds")),
         fecha,
         datos.value(QStringLiteral("fechaProgramada")),
         datos.value(QStringLiteral("plataforma")),
         datos.value(QStringLiteral("notas"))});
    const QVariant nuevoId = q.lastInsertId();
    if (!nuevoId.isValid())
        return -1;
    emit cambiado();
    return nuevoId.toInt();
}

bool RedesService::actualizar(const QVariantMap &datos)
{
    if (!datos.contains(QStringLiteral("id")))
        return false;
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("UPDATE posts_redes SET titulo = ?, contenido = ?, emojis = ?, "
                       "hashtags = ?, tipo = ?, foto_ids = ?, plataforma = ?, notas = ? WHERE id = ?"),
        {datos.value(QStringLiteral("titulo")),
         datos.value(QStringLiteral("contenido")),
         datos.value(QStringLiteral("emojis")),
         datos.value(QStringLiteral("hashtags")),
         datos.value(QStringLiteral("tipo")),
         datos.value(QStringLiteral("fotoIds")),
         datos.value(QStringLiteral("plataforma")),
         datos.value(QStringLiteral("notas")),
         datos.value(QStringLiteral("id"))});
    const bool ok = q.numRowsAffected() > 0;
    if (ok)
        emit cambiado();
    return ok;
}

bool RedesService::eliminar(int id)
{
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("DELETE FROM posts_redes WHERE id = ?"), {id});
    const bool ok = q.numRowsAffected() > 0;
    if (ok)
        emit cambiado();
    return ok;
}

bool RedesService::marcarPublicado(int id)
{
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("UPDATE posts_redes SET publicado = 1 WHERE id = ?"), {id});
    const bool ok = q.numRowsAffected() > 0;
    if (ok)
        emit cambiado();
    return ok;
}

bool RedesService::aumentarVisualizaciones(int id)
{
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("UPDATE posts_redes SET visualizaciones = visualizaciones + 1 WHERE id = ?"),
        {id});
    const bool ok = q.numRowsAffected() > 0;
    if (ok)
        emit cambiado();
    return ok;
}

QVariantMap RedesService::estadisticas() const
{
    const QVariantList todos = obtenerTodos();
    int publicados = 0, totalVis = 0;
    QMap<QString, int> porTipo, porPlataforma;
    for (const QVariant &v : todos) {
        const QVariantMap m = v.toMap();
        if (m.value(QStringLiteral("publicado")).toInt() == 1)
            ++publicados;
        totalVis += m.value(QStringLiteral("visualizaciones")).toInt();
        porTipo[m.value(QStringLiteral("tipo")).toString()]++;
        porPlataforma[m.value(QStringLiteral("plataforma")).toString()]++;
    }
    auto aLista = [](const QMap<QString, int> &m) {
        QVariantList l;
        for (auto it = m.constBegin(); it != m.constEnd(); ++it)
            l.append(QVariantMap{{QStringLiteral("nombre"), it.key()},
                                 {QStringLiteral("total"), it.value()}});
        return l;
    };
    QVariantMap out;
    out.insert(QStringLiteral("totalPosts"), todos.size());
    out.insert(QStringLiteral("publicados"), publicados);
    out.insert(QStringLiteral("noPublicados"), int(todos.size()) - publicados);
    out.insert(QStringLiteral("totalVisualizaciones"), totalVis);
    out.insert(QStringLiteral("porTipo"), aLista(porTipo));
    out.insert(QStringLiteral("porPlataforma"), aLista(porPlataforma));
    return out;
}

QString RedesService::contenidoFormateado(const QVariantMap &post) const
{
    QString out = post.value(QStringLiteral("contenido")).toString();
    const QString emojis = post.value(QStringLiteral("emojis")).toString();
    const QString hashtags = post.value(QStringLiteral("hashtags")).toString();
    if (!emojis.isEmpty())
        out += QStringLiteral("\n\n") + emojis;
    if (!hashtags.isEmpty())
        out += QStringLiteral("\n\n") + hashtags;
    return out;
}

QStringList RedesService::sugerenciasHashtags(const QString &contenido) const
{
    const QString c = contenido.toLower();
    QStringList s;
    if (c.contains(QStringLiteral("descuento")) || c.contains(QStringLiteral("oferta")))
        s << QStringLiteral("#oferta") << QStringLiteral("#descuento") << QStringLiteral("#promocion");
    if (c.contains(QStringLiteral("trabajo")) || c.contains(QStringLiteral("diseño")))
        s << QStringLiteral("#diseño") << QStringLiteral("#trabajo") << QStringLiteral("#nailart");
    s << QStringLiteral("#manicura") << QStringLiteral("#belleza")
      << QStringLiteral("#nails") << QStringLiteral("#uñas");
    // dedup preservando orden
    QStringList out;
    for (const QString &h : s)
        if (!out.contains(h))
            out << h;
    return out;
}

QStringList RedesService::sugerenciasEmojis(const QString &tipo) const
{
    if (tipo == QStringLiteral("oferta")) return {"🎉", "💰", "✨", "🎁"};
    if (tipo == QStringLiteral("promocion")) return {"💖", "🌹", "👑", "💅"};
    if (tipo == QStringLiteral("trabajo")) return {"✨", "💫", "🌟", "👌"};
    if (tipo == QStringLiteral("testimonio")) return {"💬", "⭐", "😍", "👍"};
    if (tipo == QStringLiteral("educativo")) return {"📚", "💡", "📖", "✏️"};
    return {"✨", "💅", "💖"};
}

void RedesService::copiar(const QString &texto) const
{
    if (auto *cb = QGuiApplication::clipboard())
        cb->setText(texto);
}

File: universal/lib/services/redes_service.dart
import '../config/business_config.dart';
import '../database/database_helper.dart';
import '../models/post_redes.dart';

class RedesService {
  final DatabaseHelper _db = DatabaseHelper();

  // Crear post
  Future<int> crearPost(PostRedes post) async {
    return await _db.insertPostRedes(post.toMap());
  }

  // Obtener todos los posts
  Future<List<PostRedes>> obtenerTodos() async {
    final mapList = await _db.getAllPostsRedes();
    return mapList.map((map) => PostRedes.fromMap(map)).toList();
  }

  // Obtener posts no publicados
  Future<List<PostRedes>> obtenerNoPublicados() async {
    final mapList = await _db.getPostsRedesNoPublicados();
    return mapList.map((map) => PostRedes.fromMap(map)).toList();
  }

  // Obtener posts por tipo
  Future<List<PostRedes>> obtenerPorTipo(String tipo) async {
    final todos = await obtenerTodos();
    return todos.where((p) => p.tipo == tipo).toList();
  }

  // Obtener posts por plataforma
  Future<List<PostRedes>> obtenerPorPlataforma(String plataforma) async {
    final todos = await obtenerTodos();
    return todos.where((p) => p.plataforma == plataforma).toList();
  }

  // Actualizar post
  Future<int> actualizar(PostRedes post) async {
    return await _db.updatePostRedes(post.toMap());
  }

  // Eliminar post
  Future<int> eliminar(int id) async {
    return _db.deletePostRedes(id);
  }

  // Marcar como publicado
  Future<int> marcarPublicado(int id) async {
    final posts = await obtenerTodos();
    final post = posts.firstWhere((p) => p.id == id);
    final postActualizado = post.copyWith(publicado: true);
    return await actualizar(postActualizado);
  }

  // Aumentar visualizaciones
  Future<int> aumentarVisualizaciones(int id) async {
    final posts = await obtenerTodos();
    final post = posts.firstWhere((p) => p.id == id);
    final postActualizado =
        post.copyWith(visualizaciones: post.visualizaciones + 1);
    return await actualizar(postActualizado);
  }

  // Obtener total de posts
  Future<int> totalPosts() async {
    final posts = await obtenerTodos();
    return posts.length;
  }

  // Obtener total de posts publicados
  Future<int> totalPublicados() async {
    final posts = await obtenerTodos();
    return posts.where((p) => p.publicado).length;
  }

  // Obtener posts de esta semana
  Future<List<PostRedes>> obtenerSemana() async {
    final hoy = DateTime.now();
    final hace7dias = hoy.subtract(const Duration(days: 7));
    final posts = await obtenerTodos();

    return posts
        .where((p) => p.fechaCreacion.isAfter(hace7dias))
        .toList();
  }

  // Obtener posts de este mes
  Future<List<PostRedes>> obtenerMes() async {
    final hoy = DateTime.now();
    final hace30dias = hoy.subtract(const Duration(days: 30));
    final posts = await obtenerTodos();

    return posts
        .where((p) => p.fechaCreacion.isAfter(hace30dias))
        .toList();
  }

  // Estadísticas de redes
  Future<Map<String, dynamic>> estadisticas() async {
    final todos = await obtenerTodos();
    final publicados = todos.where((p) => p.publicado).toList();
    final noPublicados = todos.where((p) => !p.publicado).toList();

    // Contar por tipo
    final porTipo = <String, int>{};
    for (var post in todos) {
      porTipo.update(
        post.tipo,
        (value) => value + 1,
        ifAbsent: () => 1,
      );
    }

    // Contar por plataforma
    final porPlataforma = <String, int>{};
    for (var post in todos) {
      porPlataforma.update(
        post.plataforma,
        (value) => value + 1,
        ifAbsent: () => 1,
      );
    }

    // Total de visualizaciones
    final totalVisualizaciones =
        todos.fold<int>(0, (sum, p) => sum + p.visualizaciones);

    return {
      'totalPosts': todos.length,
      'publicados': publicados.length,
      'noPublicados': noPublicados.length,
      'porTipo': porTipo,
      'porPlataforma': porPlataforma,
      'totalVisualizaciones': totalVisualizaciones,
      'semana': await obtenerSemana().then((l) => l.length),
      'mes': await obtenerMes().then((l) => l.length),
    };
  }

  // Exportar contenido para copiar
  String exportarParaCopiar(PostRedes post) {
    return post.getContenidoFormateado();
  }

  // Exportar lista de posts
  Future<List<Map<String, dynamic>>> exportarTodos() async {
    final posts = await obtenerTodos();
    return posts.map((p) => p.toMap()).toList();
  }

  // Generar post desde plantilla
  PostRedes generarDesdeProctilla(String tipo, String contenidoExtra) {
    final contenido = '$tipo\n\n$contenidoExtra';

    return PostRedes(
      titulo: tipo,
      contenido: contenido,
      tipo: tipo.toLowerCase(),
      plataforma: 'todas',
      fechaCreacion: DateTime.now(),
    );
  }

  // Obtener sugerencias de hashtags (según el rubro elegido)
  List<String> sugerenciasHashtags(String contenido) {
    final config = AppConfig.instance.current;
    List<String> sugerencias = [];

    if (contenido.toLowerCase().contains('descuento') ||
        contenido.toLowerCase().contains('oferta')) {
      sugerencias.addAll(['#oferta', '#descuento', '#promocion']);
    }

    if (contenido.toLowerCase().contains('trabajo') ||
        contenido.toLowerCase().contains('diseño')) {
      sugerencias.addAll(config.hashtagsComunes.take(3));
    }

    sugerencias.addAll(config.hashtagsComunes);

    return sugerencias.toSet().toList();
  }

  // Obtener sugerencias de emojis (según el rubro elegido)
  List<String> sugerenciasEmojis(String tipo) {
    final emoji = AppConfig.instance.current.emoji;
    final sugerencias = {
      'oferta': ['🎉', '💰', '✨', '🎁'],
      'promocion': ['💖', '🌹', '👑', emoji],
      'trabajo': ['✨', '💫', '🌟', '👌'],
      'testimonio': ['💬', '⭐', '😍', '👍'],
      'educativo': ['📚', '💡', '📖', '✏️'],
    };

    return sugerencias[tipo] ?? ['✨', emoji, '💖'];
  }
}
